plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -417,11 +417,88 @@ class producer_plugin_impl : public std::enable_shared_from_this<producer_plugin
          packed_transaction_ptr  trx;
          next_func_t             next;
       };
+   
       // The queue storing read-only transactions to be executed by read-only threads
-      struct ro_trx_queue_t {
-         std::mutex              mtx;
-         std::deque<ro_trx_t>    queue;
+      class ro_trx_queue_t {
+      public:
+         void push_back(ro_trx_t&& trx) {
+            std::unique_lock<std::mutex> g( mtx );
+            queue.push_back(std::move(trx));
+            if (num_waiting)
+               cond.notify_one();
+         }
+         
+         void push_front(ro_trx_t&& trx) {
+            std::unique_lock<std::mutex> g( mtx );
+            queue.push_front(std::move(trx));
+            if (num_waiting)
+               cond.notify_one();
+         }
+
+         bool empty() const {
+            std::lock_guard<std::mutex> g( mtx );
+            return queue.empty();
+         }
+
+         // may wait if the queue is empty, and not all other threads are already waiting.
+         // returns true if a transaction was dequeued and should be executed, or false 
+         // if conditions are met to stop processing transactions. 
+         bool pop_front(ro_trx_t &trx) {
+            std::unique_lock<std::mutex> g( mtx );
+            
+            if (should_exit()) {
+               cond.notify_all();
+               return false;
+            }
+
+            if (queue.empty()) {
+               if (num_waiting + 1 == max_waiting) {
+                  exiting_read_window = true;
+                  cond.notify_all();
+                  return false;
+               }
+               
+               ++num_waiting;
+               cond.wait(g, [this]() { return !queue.empty() || should_exit(); });
+               --num_waiting;
+               
+               if (should_exit())
+                  return false;
+            }
+            
+            trx = std::move(queue.front());
+            queue.pop_front();
+            return true;
+         }
+
+         // We exit the read window when either:
+         //    - all threads would be idle
+         //    - or the net_plugin received a block.
+         //    - or we have reached the read_window_deadline
+         void set_exit_criteria(uint32_t num_tasks, std::atomic<bool>* received_block, fc::time_point deadline) {
+            std::lock_guard<std::mutex> g( mtx ); // not strictly necessary with current usage from single thread
+            assert(num_tasks > 0 && num_waiting == 0 && received_block != nullptr);
+            assert(received_block && *received_block == false);
+            max_waiting = num_tasks;
+            received_block_ptr = received_block;
+            read_window_deadline = deadline;
+         }
+
+      private:
+         bool should_exit() {
+            return *received_block_ptr || exiting_read_window || fc::time_point::now() >= read_window_deadline;
```

Comment:
I would check `*received_block_ptr` last as it is more expensive to check than `exiting_read_window`.

Reply:
done

Reply:
> as it is more expensive

Because of the `atomic`? It is pronbably more expensive than `exiting_read_window`, but less than `fc::time_point::now()`